### ISSUE TITLE
Weak Hash APIs and Insecure Random Functions 

### ISSUE THREAD

**Issue by agnihotriayush** [aws-amplify/aws-sdk-android#3017]
**Which AWS Services is the feature request for?**
com.amazonaws:aws-android-sdk-s3:2.7.+

**Is your feature request related to a problem? Please describe.**
The android application was found to use various dangerous and insecure functions as well as APIs. The following affected APIs and functions are listed below:
Weak hash APIs: AWS was found to be using MD5. These standards are no longer regarded as secure.
Insecure Random Functions: AWS was found to be using java.util.Random. Use of these functions is not cryptographically secure for generating pseudorandom numbers.

Weak Hash APIs 	
o	com/amazonaws/services/s3/internal/MD5DigestCalculatingInputStream.java
o	com/amazonaws/services/s3/AmazonS3Client.java
o	com/amazonaws/util/Md5Utils.java 
Insecure Random Functions 
o 	com/amazonaws/retry/PredefinedRetryPolicies.java 

**Describe the solution you'd like**
Investigate whether or not these dangerous functions are employed in a way where buffer overflow conditions will be allowed. Switch to alternate and secure functions and stronger hashing algorithms

**Comment by gpanshu** [comment]
Thank you for posting this question. We are looking into this and will reply soon with our findings.

**Comment by eeatonaws** [comment]
Hello, and thank you for your patience while we investigated your reported concerns. We have analyzed the reported concerns, and they are not issues impacting the AWS SDK for Android. A more detailed response for each reported concern is provided below.
​

Weak Hash APIs:

MD5 is used to generate the value for the "Content-MD5" request parameter for multiple S3 requests. The "Content-MD5" request parameter is an MD5 value used to proactively verify the integrity of an object uploaded to S3. When you upload objects to Amazon S3, and pass the Content-MD5 value as a request header, Amazon S3 checks the object against the provided Content-MD5 value. If the values do not match, you receive an error. The usage of MD5 in this context does not represent an issue or concern, and MD5 is not used within the SDK for any security-related operations.


Insecure Random Functions:

The random function is only used to calculate “jitter” in network request retries (https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/), and not for any cryptographically sensitive operations that require true randomness. The usage of random in this context does not represent an issue.
